compiler/src/test/scala/org/scalajs/core/compiler/test/SJSDefinedByDefaultTest.scala

```diff
@@ -0,0 +1,246 @@
+      |                   ^
+    """
+  }
+
+  @Test
+  def noOverrideConcreteNonOptionalWithOptionalSeparateCompilation1: Unit = {
+    """
+    abstract class A extends js.Object {
+      val a1: js.UndefOr[Int] = 5
+      val a2: js.UndefOr[Int]
+
+      def b1: js.UndefOr[Int] = 5
+      def b2: js.UndefOr[Int]
+    }
+    """.succeeds()
+
+    """
+    trait B extends A {
+      override val a1: js.UndefOr[Int] = js.undefined
+      override val a2: js.UndefOr[Int] = js.undefined
+
+      override def b1: js.UndefOr[Int] = js.undefined
+      override def b2: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:6: error: Cannot override concrete val a1: scala.scalajs.js.UndefOr[Int] from A in a Scala.js-defined JS trait.
+      |      override val a1: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:9: error: Cannot override concrete def b1: scala.scalajs.js.UndefOr[Int] from A in a Scala.js-defined JS trait.
+      |      override def b1: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
+  }
+
+  @Test
+  def noOverrideConcreteNonOptionalWithOptionalSeparateCompilation2: Unit = {
+    """
+    @js.native
+    @JSGlobal
+    class A extends js.Object {
+      val a: js.UndefOr[Int] = js.native
+      def b: js.UndefOr[Int] = js.native
+    }
+    """.succeeds()
+
+    """
+    trait B extends A {
+      override val a: js.UndefOr[Int] = js.undefined
+      override def b: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:6: error: Cannot override concrete val a: scala.scalajs.js.UndefOr[Int] from A in a Scala.js-defined JS trait.
+      |      override val a: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:7: error: Cannot override concrete def b: scala.scalajs.js.UndefOr[Int] from A in a Scala.js-defined JS trait.
+      |      override def b: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
+  }
+
+  @Test
+  def noOverrideConcreteNonOptionalWithOptionalSeparateCompilation3: Unit = {
+    """
+    @js.native
+    trait A extends js.Object {
+      val a: js.UndefOr[Int] = js.native
+      def b: js.UndefOr[Int] = js.native
+    }
+
+    @js.native
+    @JSGlobal
+    class B extends A
+    """.succeeds()
+
+    """
+    trait C extends B {
+      override val a: js.UndefOr[Int] = js.undefined
+      override def b: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:6: error: Cannot override concrete val a: scala.scalajs.js.UndefOr[Int] from A in a Scala.js-defined JS trait.
+      |      override val a: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:7: error: Cannot override concrete def b: scala.scalajs.js.UndefOr[Int] from A in a Scala.js-defined JS trait.
+      |      override def b: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
+  }
+
+}
```

Comment:
Can you also add a test that a Scala.js defined class cannot directly extend a native JS trait? Just for extra safety :)